### PR TITLE
chore(repo): remove useDeepCompareEffect where possible

### LINKED ISSUE #1797: [repo] useDeepCompareEffect, dequal - RangeError: Maximum call stack size exceeded
### What package is this for?

- [x] React
- [ ] Angular

### Describe the bug

`useDeepCompareEffect` [does not support](https://github.com/kentcdodds/use-deep-compare-effect/issues/7) comparing props (or prop objects/arrays that contain props) of type Element or Node, `useDeepCompareEffect` is unable to compare them due to these including the entire (very large) React tree.

This presents as an error of the underlying comparison library within `useDeepCompareEffect`, `dequal` in both browser and/or tests:

```
RangeError: Maximum call stack size exceeded

      at dequal (node_modules/dequal/dist/index.js:9:16)
      at dequal (node_modules/dequal/dist/index.js:77:28)
      at dequal (node_modules/dequal/dist/index.js:77:28)
      at ...
```

One way to partially fix this is to stop blindly comparing huge prop objects, like `table.view` for instance, and instead compare the specific props within that are not of type Element or Node. I've submitted a PR for this for StatefulTable/Table https://github.com/carbon-design-system/carbon-addons-iot-react/pull/1782 but we have additional components that we need to investigate

- [x] [`CardRenderer.jsx`](https://github.com/carbon-design-system/carbon-addons-iot-react/blob/d63c51871be9006f698354af6821d5cbdb637294/src/components/Dashboard/CardRenderer.jsx#L31)
- [x] [`ImageHotspots.jsx`](https://github.com/carbon-design-system/carbon-addons-iot-react/blob/1644c8407ea05e7cdcb7900c91f847ba663a3904/src/components/ImageCard/ImageHotspots.jsx#L365)
- [x] [`HierarchyList.jsx`](https://github.com/carbon-design-system/carbon-addons-iot-react/blob/581e8e247530a6979354de727c6694285d0fd0b4/src/components/List/HierarchyList/HierarchyList.jsx#L209) (this is confirmed to be comparing Nodes within the `items` prop and is prone to this error)
- [x] [`TableHead.jsx`](https://github.com/carbon-design-system/carbon-addons-iot-react/blob/2aac0dc996a2f361cc05c0abcea87292a0c50b69/src/components/Table/TableHead/TableHead.jsx#L285)
- [x] [`StatefulTtileCatalog.jsx`](https://github.com/carbon-design-system/carbon-addons-iot-react/blob/1644c8407ea05e7cdcb7900c91f847ba663a3904/src/components/TileCatalog/StatefulTileCatalog.jsx#L39)
- [x] [`TimeSeriesCard.jsx`](https://github.com/carbon-design-system/carbon-addons-iot-react/blob/48f9a118e9e8d36e74ec0ecf2455afc93bff3b3c/src/components/TimeSeriesCard/TimeSeriesCard.jsx#L424)

This is just an initial list that will become outdated - search for usages of `useDeepCompareEffect` to find a current list.

### To Reproduce

This is only reproducible in environments with babel deps versions >=7.12.0, which is why we don't see the issue currently our own codebase/storybook.

### Expected behavior

[Deep comparison is an anti-pattern](https://twitter.com/dan_abramov/status/1104414272753487872). Most particularly with objects of unknown size like the nested structure of `items` for HierarchyList.

#### Ideal behavior:
- Components don't use `useDeepCompareEffect` 

#### Not ideal, but acceptable behavior:
- Components that absolutely require usage of `useDeepCompareEffect` depend on more specific sub props, do not compare Nodes or Elements, and thoroughly document why it's required and a normal `useEffect` doesn't work.

### Environment/versions:

- `carbon-addons-iot-react` version: 2.140.0

### PR DESCRIPTION
**Pull request by kevinsperrine**
Closes #1797 

**Summary**

- Went through each component and replaced useDeepCompareEffect with useEffect where possible.

**Change List (commits, features, bugs, etc)**

- added a [usePrevious hook](https://reactjs.org/docs/hooks-faq.html#how-to-get-the-previous-props-or-state)
- CardRenderer: was using useDeepCompareEffect to compare the styles. I replaced it with a usePrevious hook and do the comparison inside useEffect with lodash isEqual.
- DashboardEditorCardRenderer: same as above
- ImageHotspots: was using useDeepCompareEffect to trigger zooming when parameters changed. I split all the necessary dependencies out into their individual parts and switched it to useEffect
- ImageGalleryModal: was using useDeepCompareEffect to compare the content prop. Again, I replaced it with a usePrevious hook and do the comparison inside useEffect with lodash isEqual
- HierarchyList: used useDeepCompareEffect to compare items and their children when filtering by value. I added a helper method to strip these complex items down to their basics needed for this comparison `{id, children}` and then memoize that result and use it in the useEffect instead of the full items to prevent nodes from the icon and/or tags being compared and triggering the memory error.
- StatefulTileCatalog: used useDeepCompareEffect to deeply compare the tile prop. Replaced with a usePrevious hook and do the comparison inside useEffect with lodash isEqual.
- TimeSeriesCard: used deep compare effect on chart data components to generate and set that data on the chart model. I switched the order to use the memoized chartData as generated by `formatChartData`, save it in a usePrevious hook, and do the comparison inside useEffect using lodash isEqual to determine if the chart data modal needs updated or not.
- Table/StatefulTable: still uses useDeepCompareEffect, because there's so much going on there I didn't touch it.

**Acceptance Test (how to verify the PR)**

- All tests pass
- Stories for all these components still match functionally to current next branch.

**Regression Test (how to make sure this PR doesn't break old functionality)**

- Stories for all these components still match functionally to current next branch.

### CODE REVIEW

File: packages/react/src/components/List/HierarchyList/HierarchyList.jsx [hunk 1]
```diff
@@ -233,18 +247,21 @@ const HierarchyList = ({
...
-    } else if (selectedIds.includes(id) && hasDeselection) {
-      setSelectedIds(selectedIds.filter((item) => item !== id));
-      // else, no-op because the item can't be deselected
-    } else if (hasMultiSelect) {
-      setSelectedIds([...selectedIds, id]);
-    } else {
-      setSelectedIds([id]);
-    }
-  };
+  const setSelected = useCallback(
```
[the remark this answers is not in the document]
Comment (kevinsperrine) [reply packages/react/src/components/List/HierarchyList/HierarchyList.jsx:250]: that was auto-added by my editor at some point. I'll strip that back out.

File: packages/react/src/components/List/HierarchyList/HierarchyList.jsx [hunk 2]
```diff
@@ -218,10 +229,13 @@ const HierarchyList = ({
   const [selectedIds, setSelectedIds] = useState([]);
   const [editModeSelectedIds, setEditModeSelectedIds] = useState([]);
   const [showModal, setShowModal] = useState(false);
-
-  useDeepCompareEffect(() => {
-    setFilteredItems(items);
-  }, [items]);
+  const itemsWithIdAndChildren = useMemo(() => reduceItems(items), [items]);
```
[the remark this answers is not in the document]
Comment (kevinsperrine) [reply packages/react/src/components/List/HierarchyList/HierarchyList.jsx:232]: sure thing.

File: packages/react/src/components/TileCatalog/StatefulTileCatalog.jsx [hunk 3]
```diff
@@ -32,24 +34,37 @@ const StatefulTileCatalog = ({
...
+          },
+        });
+        // If we totally change the tiles data, we should generate a selection event for the initial default selection
+        if (isSelectedByDefault && onSelection && tilesProp.length > 0 && !selectedTileIdProp) {
+          onSelection(tilesProp[0].id);
+        }
+      }
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [tilesWithoutRenderContent]
```
[the remark this answers is not in the document]
Comment (kevinsperrine) [reply packages/react/src/components/TileCatalog/StatefulTileCatalog.jsx:66]: good catch. forgot to add it before adding the eslint disable comment.

File: packages/react/src/hooks/usePrevious.jsx [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import { useEffect, useRef } from 'react';
+
+export function usePrevious(value) {
```
[the remark this answers is not in the document]
Comment (kevinsperrine) [reply packages/react/src/hooks/usePrevious.jsx:3]: absolutely.

**Review by sstone2423 — APPROVED** [review]
Nice!

**Pull request merged**